### DISCUSSION TITLE
Implementing service in another file

### DISCUSSION THREAD

**Discussion by theo-m** [hyperium/tonic#828] [Q&A] [answered]
Hello, 
I'm a rust beginner - went through the rustlings exercises and I'm working on my first rust project!
I went through the tutorial, and since I'm working on a well specified project I wanted to start doing things "right". For me that meant implementing my `UserService` in a dedicated file (hence module (?)).
So I'd have this setup:
<details>

```rust
# main.rs
use tonic::transport::Server;

use pb::user_service_server::UserServiceServer;
use pb::user_service_server::UserService;
use user::UserServiceImpl;

mod user;

mod pb {
    tonic::include_proto!("app");
}


#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let addr = "[::1]:50051".parse()?;
    let user_service = UserServiceImpl::default();

    Server::builder()
        .add_service(UserServiceServer::new(user_service))
        .serve(addr)
        .await?;

    Ok(())
}
``` 
and next to it another file responsible for the implementation of the `pb::user_service_server::UserServiceServer`:
```rust
# user.rs
use tonic::{Request, Response, Status};

mod pb {
    tonic::include_proto!("app");
}

#[derive(Debug, Default)]
pub struct UserServiceImpl {}


#[tonic::async_trait]
impl pb::user_service_server::UserService for UserServiceImpl {
    async fn sync_contacts(
        &self,
        request: Request<pb::SyncContactsInput>,
    ) -> Result<Response<pb::SyncContactsOutput>, Status> {
        unimplemented!()
    }

    async fn onboarding(
        &self,
        request: Request<pb::OnboardingInput>,
    ) -> Result<Response<pb::MeUser>, Status> {
        unimplemented!()
    }

    async fn me(
        &self,
        request: Request<pb::Empty>,
    ) -> Result<Response<pb::MeUser>, Status> {
        unimplemented!()
    }

    async fn get(
        &self,
        request: Request<pb::UserGetInput>,
    ) -> Result<Response<pb::User>, Status> {
        unimplemented!()
    }
}
```

</details>

But compiling fails with: 

> the trait `pb::user_service_server::UserService` is not implemented for `UserServiceImpl`

I've been looking around and saw different solutions on stackoverflow but none made my program compile. In the current state concatening the content of the two files is a compiling program.
Quite randomly stumbled onto [this article about the "orphan rule"](https://blog.mgattozzi.dev/orphan-rules/) and it seems like this is the problem I'm hitting?

Could not find examples of `tonic` use with services split across different modules and files - maybe it's a good addition to the docs?

**Comment by theo-m** [comment, marked as answer]
I've received pointers from elsewhere:
The solution is to `include_proto!` only once, e.g. for me 
```rust
// main.rs
use tonic::transport::Server;

use pb::user_service_server::UserServiceServer;
use pb::user_service_server::UserService;
use user::UserServiceImpl;

mod user;
use user::pb;

// don't do this!
// mod pb {
//    tonic::include_proto!("app");
// }
```

```rust
// user.rs
use tonic::{Request, Response, Status};

// v--- make the thing public! ---v
pub mod pb {
    tonic::include_proto!("app");
}

#[derive(Debug, Default)]
pub struct UserServiceImpl {}
```

**Reply by Ben2146053** [reply]
Thanks for this post, I had the same problem.
